File: anvil/ui/widgets/temp_gauge.py
"""
TempGauge — Temperature display with a LevelBar.
"""
import gi
gi.require_version("Gtk", "4.0")
gi.require_version("Adw", "1")
from gi.repository import Gtk, Adw, Gdk  # noqa: E402


def _pick_icon_name(*candidates: str) -> str:
    """Return the first icon that exists in the active icon theme."""
    display = Gdk.Display.get_default()
    if display:
        icon_theme = Gtk.IconTheme.get_for_display(display)
        for name in candidates:
            if icon_theme.has_icon(name):
                return name
    return candidates[-1] if candidates else "image-missing"


class TempGauge(Adw.ActionRow):
    """ActionRow showing CPU temperature with a coloured LevelBar."""

    def __init__(self):
        super().__init__(title="CPU Temperature")
        self.set_activatable(False)
        self.add_prefix(
            Gtk.Image.new_from_icon_name(
                _pick_icon_name(
                    "sensors-temperature-symbolic",
                    "temperature-high-symbolic",
                    "weather-clear-symbolic",
                    "dialog-information-symbolic",
                )
            )
        )

        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=8)
        box.set_valign(Gtk.Align.CENTER)

        # Numeric label
        self._label = Gtk.Label(label="—", css_classes=["dim-label"])
        self._label.set_width_chars(5)
        self._label.set_xalign(1.0)
        box.append(self._label)

        # Level bar
        self._bar = Gtk.LevelBar()
        self._bar.set_min_value(0)
        self._bar.set_max_value(105)
        self._bar.set_value(0)
        self._bar.set_hexpand(False)
        self._bar.set_size_request(120, -1)
        self._bar.set_valign(Gtk.Align.CENTER)
        self._bar.add_css_class("temp-gauge")

        # Named offsets for colour thresholds
        self._bar.add_offset_value("low", 55)
        self._bar.add_offset_value("high", 75)
        self._bar.add_offset_value("full", 90)

        box.append(self._bar)

        self.add_suffix(box)

    def set_temperature(self, temp_c: float | None):
        """Update the gauge with a new temperature reading."""
        if temp_c is None:
            self._label.set_label("N/A")
            self._bar.set_value(0)
            return

        self._label.set_label(f"{temp_c:.0f} °C")
        self._bar.set_value(min(temp_c, 105))

File: src/ui/widgets/temp_gauge.rs
use gtk4::prelude::*;
use libadwaita::prelude::*;

fn pick_icon_name(candidates: &[&str]) -> String {
    if let Some(display) = gtk4::gdk::Display::default() {
        let icon_theme = gtk4::IconTheme::for_display(&display);
        for name in candidates {
            if icon_theme.has_icon(name) {
                return name.to_string();
            }
        }
    }
    candidates
        .last()
        .unwrap_or(&"dialog-information-symbolic")
        .to_string()
}

pub struct TempGauge {
    pub row: libadwaita::ActionRow,
    label: gtk4::Label,
    bar: gtk4::LevelBar,
}

impl TempGauge {
    pub fn new() -> Self {
        let row = libadwaita::ActionRow::new();
        row.set_title("CPU Temperature");
        row.set_activatable(false);

        let icon_name = pick_icon_name(&[
            "thermometer-symbolic",
            "sensors-temperature-symbolic",
            "temperature-high-symbolic",
            "weather-clear-symbolic",
            "dialog-information-symbolic",
        ]);
        row.add_prefix(&gtk4::Image::from_icon_name(&icon_name));

        let box_ = gtk4::Box::new(gtk4::Orientation::Horizontal, 8);
        box_.set_valign(gtk4::Align::Center);

        let label = gtk4::Label::new(Some("—"));
        label.add_css_class("dim-label");
        label.set_width_chars(5);
        label.set_xalign(1.0);
        box_.append(&label);

        let bar = gtk4::LevelBar::new();
        bar.set_min_value(0.0);
        bar.set_max_value(105.0);
        bar.set_value(0.0);
        bar.set_hexpand(false);
        bar.set_size_request(120, -1);
        bar.set_valign(gtk4::Align::Center);
        bar.add_css_class("temp-gauge");
        bar.add_offset_value("low", 55.0);
        bar.add_offset_value("high", 75.0);
        bar.add_offset_value("full", 90.0);
        box_.append(&bar);

        row.add_suffix(&box_);

        TempGauge { row, label, bar }
    }

    pub fn set_temperature(&self, temp_c: Option<f64>) {
        match temp_c {
            None => {
                self.label.set_label("N/A");
                self.bar.set_value(0.0);
            }
            Some(temp) => {
                self.label.set_label(&format!("{:.0} °C", temp));
                self.bar.set_value(temp.min(105.0));
            }
        }
    }

    pub fn widget(&self) -> &libadwaita::ActionRow {
        &self.row
    }
}
